src/serializers/USDSerializer.cpp

```diff
@@ -0,0 +1,166 @@
+
+		if(materials_.find(material_path) != materials_.end()) {
+			materials.push_back(materials_[material_path]);
+			continue;
+		}
+
+  	  	const std::string path("/Looks/" + material_path);
+  	  	auto material = pxr::UsdShadeMaterial::Define(stage_, pxr::SdfPath(path));
+  	  	auto shader = pxr::UsdShadeShader::Define(stage_, pxr::SdfPath(path + "/Shader"));
+  	  	shader.CreateIdAttr().Set(pxr::TfToken("UsdPreviewSurface"));
+
+  	  	float rgba[4] { 0.18f, 0.18f, 0.18f, 1.0f };
+  	  	if (style.hasDiffuse())
+			for (int i = 0; i < 3; ++i)
+  	  	    	rgba[i] = static_cast<float>(style.diffuse()[i]);
+  	  	shader.CreateInput(pxr::TfToken("diffuseColor"), pxr::SdfValueTypeNames->Color3f).Set(pxr::GfVec3f(rgba[0], rgba[1], rgba[2]));
+
+  	  	if(style.hasTransparency())
+  	  		rgba[3] -= style.transparency();
+  	  	shader.CreateInput(pxr::TfToken("opacity"), pxr::SdfValueTypeNames->Float).Set(rgba[3]);
+
+  	  	if(style.hasSpecular()) {
+  	  		for (int i = 0; i < 3; ++i)
+  	  	    	rgba[i] = static_cast<float>(style.specular()[i]);
+  	  	  	shader.CreateInput(pxr::TfToken("useSpecularWorkflow"), pxr::SdfValueTypeNames->Int).Set(1);
+  	  	} else {
+  	  	  	shader.CreateInput(pxr::TfToken("useSpecularWorkflow"), pxr::SdfValueTypeNames->Int).Set(0);
+  	  	}
+  	  	shader.CreateInput(pxr::TfToken("specularColor"), pxr::SdfValueTypeNames->Color3f).Set(pxr::GfVec3f(rgba[0], rgba[1], rgba[2]));
+
+  	  	material.CreateSurfaceOutput().ConnectToSource(shader.ConnectableAPI(), pxr::TfToken("surface"));
+  	  	materials.push_back(material);
+		materials_[material_path] = material;
+  	}
+
+  	return materials;
+}
+
+void USDSerializer::writeHeader() {
+  	stage_->GetRootLayer()->SetComment("File generated by IfcOpenShell " + std::string(IFCOPENSHELL_VERSION)); 
+}
+
+void USDSerializer::write(const IfcGeom::TriangulationElement* o) {
+  	const IfcGeom::Representation::Triangulation& mesh = o->geometry();
+  	const auto verts = mesh.verts();
+  	const auto faces = mesh.faces();
+  	const auto material_ids = mesh.material_ids();
+  	const std::vector<double>& m = o->transformation().matrix().data();
+
+	if (material_ids.empty() || verts.empty() || faces.empty())
+		return;
+
+	std::string name = o->name();
+  	if(name.empty()) {
+  	  	name = "UnNamed";
+  	} else {
+  	  	name = usd_utils::toPath(name);
+  	}
+  	name += "_" + std::to_string(o->id());
+  	auto usd_mesh = pxr::UsdGeomMesh::Define(stage_, pxr::SdfPath("/World/" + name));
+
+  	usd_mesh.AddTranslateOp().Set(pxr::GfVec3d(m[9], m[10], m[11]));
+
+  	pxr::VtVec3fArray points;
+  	for(std::size_t i = 0; i < verts.size(); i+=3) {
+  	  	points.push_back(pxr::GfVec3f(static_cast<float>(verts[i] * m[0] + verts[i+1] * m[3] + verts[i+2] * m[6]),
+  	                                  static_cast<float>(verts[i] * m[1] + verts[i+1] * m[4] + verts[i+2] * m[7]),
+  	                                  static_cast<float>(verts[i] * m[2] + verts[i+1] * m[5] + verts[i+2] * m[8])));
+  	}
```

Comment:
Any particular reason why you choose to use a translation, but factor in the rotation into the vertex coords? That way you cannot reuse meshes across different instances (assuming usd has a concept like that).

`o->transformation().matrix().data()` comes from an https://ifc43-docs.standards.buildingsmart.org/IFC/RELEASE/IFC4x3/HTML/lexical/IfcAxis2Placement3D.htm so it's guaranteed to be only rotation + translation, no scaling or shear.

Otherwise, an other option is to use USE_WORLD_COORDS in the settings. That way you can just ignore the matrix altogether and have ifcopenshell factor it into the coords for you.

Reply:
(Objects in IFC sharing the same representation will be sequential in the iterator output and have the same `o->geometry().id()`)

Reply:
You are right! I'm just getting to know USD and it's concepts. I know there is support for referencing etc, just have to figure out how it works